courses/intro-to-ada/contracts.rst

```diff
@@ -0,0 +1,485 @@
+won't be changed by ``Op``. In other words, the caller expects that ``Op``
+doesn't modify the argument it's providing, but rather just read the
+information stored in the argument. Constraints and subtypes are other
+examples of contracts. In summary, these elements allow for improving the
+consistency of the application.
+
+In general, design-by-contract programming refers to techniques that
+include pre- and postconditions, type invariants, and subtype predicates.
+We will look into those topics in this section.
+
+Pre and postconditions
+----------------------
+
+Pre and postconditions refer to expectations regarding input and output
+parameters of subprograms, and return value of functions. If we say that
+certain requirements should be met before calling a subprogram ``Op``,
+we're talking about preconditions. Similarly, if certain requirements
+should be met after a call to the subprogram ``Op``, we're talking about
+postconditions.
+
+Pre and postconditions are specified by using a :ada:`with` clause in the
+subprogram declaration. A :ada:`with Pre => <condition>` clause
+specifies a precondition, whereas a :ada:`with Post => <condition>` clause
+specifies a postcondition.
+
+The following code shows an example of preconditions:
+
+.. code:: ada
+    :class: ada-run-expect-failure
+
+    with Ada.Text_IO; use Ada.Text_IO;
+    with Ada.Numerics.Elementary_Functions;
+
+    procedure Show_Simple_Precondition is
+       pragma Assertion_Policy (Pre  => Check);
+
+       procedure DB_Entry (Name : String;
+                           Age  : Natural)
+         with Pre => Name'Length > 0
+       is
+       begin
+          --  Missing implementation
+          null;
+       end DB_Entry;
+    begin
+       DB_Entry ("John", 30);
+       DB_Entry ("",     21);  --  Precondition will fail!
+    end Show_Simple_Precondition;
+
+In this example, we want to ensure that the name field in our database
+doesn't contain an empty string. We can implement this requirement by
+using a precondition that ensures that the length of the string used for
+the :ada:`Name` parameter of the :ada:`DB_Entry` procedure is greater than
+zero. If another subprogram attempts to call the :ada:`DB_Entry` procedure
+with an empty string for the :ada:`Name` parameter, the call will fail
+because the precondition is not met.
+
+Note that the :ada:`pragma Assertion_Policy` statement is used to force
+the compiler to generate a check for the precondition. The same
+:ada:`pragma` will be used for other kinds of contracts in the remaining
+examples from this section.
+
+The following code shows an example of a postcondition:
+
+.. code:: ada
+    :class: ada-run-expect-failure
+
+    with Ada.Text_IO; use Ada.Text_IO;
+    with Ada.Numerics.Elementary_Functions;
+
+    procedure Show_Simple_Postcondition is
+       pragma Assertion_Policy (Post => Check);
+
+       type Int_8 is range -2 ** 7 .. 2 ** 7 - 1
+         with Size => 8;
```

Comment:
It's just that the type is called `Int_8`, so that using `Size` ensures that the compiler will actually create an 8-bit type --- or it will explicitly mention that the type requires a larger amount of bytes. In any case, I'll remove all instances of it because it's bothering you.